lib/src/Mailbox.cc

```diff
@@ -457,3 +458,159 @@ namespace kickcat
+
+    void Mailbox::receive()
+    {
+        std::vector<uint8_t> raw_message;
+        raw_message.resize(mbx_in_.length);
+        int32_t read_bytes = esc_->read(mbx_in_.start_address, raw_message.data(), mbx_in_.length);
+        if (read_bytes != mbx_in_.length)
+        {
+            return;
+        }
+
+        auto const* header  = pointData<mailbox::Header>(raw_message.data());
+        if ((header->type == mailbox::ERR) or (header->len == 0))
+        {
+            replyError(std::move(raw_message), mailbox::Error::INVALID_HEADER);
+            return;
+        }
+
+        for (auto it = to_process_.begin(); it != to_process_.end(); ++it)
+        {
+            ProcessingResult state = (*it)->process(raw_message);
+            switch (state)
+            {
+                case ProcessingResult::NOOP:
+                {
+                    continue;
+                }
+                case ProcessingResult::CONTINUE:
+                case ProcessingResult::FINALIZE_AND_KEEP:
+                {
+                    return;
+                }
+                case ProcessingResult::FINALIZE:
+                {
+                    it = to_process_.erase(it);
+                    return;
+                }
+                default: { }
+            }
+        }
+
+        if (to_process_.size() >= max_msgs_)
+        {
+            // Queue is full and no one process it: drop the message
+            replyError(std::move(raw_message), mailbox::Error::NO_MORE_MEMORY);
+            return;
+        }
+
+        // No message handle it: let's try to build a new one
+        for (auto& factory : factories_)
+        {
+            auto msg = factory(this, std::move(raw_message));
+            if (msg != nullptr)
+            {
+                to_process_.push_back(msg);
+                return;
+            }
+        }
+
+        // Nothing can be done with this message
+        replyError(std::move(raw_message), mailbox::Error::UNSUPPORTED_PROTOCOL);
+    }
+
+
+    void Mailbox::process()
+    {
+        for (auto it = to_process_.begin(); it != to_process_.end(); ++it)
+        {
+            ProcessingResult state = (*it)->process();
+            switch (state)
+            {
+                case ProcessingResult::NOOP:
+                {
+                    continue;
+                }
+                case ProcessingResult::CONTINUE:
+                case ProcessingResult::FINALIZE_AND_KEEP:
+                {
+                    return;
+                }
+                case ProcessingResult::FINALIZE:
+                {
+                    it = to_process_.erase(it);
+                    return;
+                }
+                default: { }
+            }
+        }
+    }
+
+    void Mailbox::send()
+    {
+        if (to_send_.empty())
+        {
+            return;
+        }
+
+        auto& msg = to_send_.front();
+        int32_t written_bytes = esc_->write(mbx_out_.start_address, msg.data(), msg.size());
```

Comment:
Is there no check to perform to know if the previous message was properly used by the ESC ?

Reply:
There is: if the message cannot be written, the message is kept in the sending list. Note that the ESC implementation shall check if the write can be done or the result of the write access.